File: app/src/androidMain/kotlin/com/thejawnpaul/gptinvestor/core/remoteconfig/RemoteConfigImpl.kt
package com.thejawnpaul.gptinvestor.core.remoteconfig

import com.google.firebase.Firebase
import com.google.firebase.remoteconfig.ConfigUpdate
import com.google.firebase.remoteconfig.ConfigUpdateListener
import com.google.firebase.remoteconfig.FirebaseRemoteConfigException
import com.google.firebase.remoteconfig.remoteConfig
import com.google.firebase.remoteconfig.remoteConfigSettings
import com.thejawnpaul.gptinvestor.BuildConfig
import com.thejawnpaul.gptinvestor.R
import com.thejawnpaul.gptinvestor.core.firebase.IRemoteConfig
import org.koin.core.annotation.Single
import timber.log.Timber
import com.google.firebase.remoteconfig.FirebaseRemoteConfig as RemoteConfig

@Single
class RemoteConfigImpl : IRemoteConfig {

    var remoteConfig: RemoteConfig = Firebase.remoteConfig

    init {
        val configSettings = if (BuildConfig.DEBUG) {
            remoteConfigSettings {
                minimumFetchIntervalInSeconds = 3600
            }
        } else {
            remoteConfigSettings { }
        }
        remoteConfig.setConfigSettingsAsync(configSettings)
        remoteConfig.setDefaultsAsync(R.xml.remote_config_defaults)

        remoteConfig.addOnConfigUpdateListener(object : ConfigUpdateListener {
            override fun onUpdate(configUpdate: ConfigUpdate) {
                Timber.e("Updated keys: %s", configUpdate.updatedKeys)
            }

            override fun onError(error: FirebaseRemoteConfigException) {
                Timber.e("Config update error with code: %s", error.code)
            }
        })
    }

    override suspend fun fetchAndActivateStringValue(configKey: String): String {
        var result = remoteConfig.getString(configKey)

        remoteConfig.fetchAndActivate().addOnCompleteListener { task ->
            if (task.isSuccessful) {
                result = remoteConfig.getString(configKey)
            }
        }
        return result
    }

    override suspend fun fetchAndActivateValue(configKey: String): Float {
        var result = remoteConfig.getDouble(configKey)

        remoteConfig.fetchAndActivate().addOnCompleteListener { task ->
            if (task.isSuccessful) {
                result = remoteConfig.getDouble(configKey)
            }
        }
        return result.toFloat()
    }

    override suspend fun fetchAndActivateBooleanValue(configKey: String): Boolean {
        var result = remoteConfig.getBoolean(configKey)

        remoteConfig.fetchAndActivate().addOnCompleteListener { task ->
            if (task.isSuccessful) {
                result = remoteConfig.getBoolean(configKey)
            }
        }
        return result
    }
}

File: iosApp/iosApp/Firebase/RemoteConfigImpl.swift
//
//  RemoteConfigImpl.swift
//  iosApp
//
//  Created by ABDULKARIM ABDULRAHMAN on 19/07/2025.
//  Copyright © 2025 orgName. All rights reserved.
//
import GPT_Investor
import FirebaseRemoteConfig

class RemoteConfigImpl: IRemoteConfig {
    
    var remoteConfig: RemoteConfig!
    init() {
        remoteConfig = RemoteConfig.remoteConfig()
        let settings = RemoteConfigSettings()
        #if DEBUG
            settings.minimumFetchInterval = 3600
        #else
            settings.minimumFetchInterval = 0
        #endif
        remoteConfig.configSettings = settings
        remoteConfig.setDefaults(fromPlist: "RemoteConfigDefaults")
        addConfigUpdateListener()
    }
    
    private func addConfigUpdateListener() {
        remoteConfig.addOnConfigUpdateListener { (configUpdate: RemoteConfigUpdate?, error: Error?) -> Void in
            guard let configUpdate = configUpdate, error == nil else { // Added an explicit unwrap for configUpdate here for clarity if you use it this way
                print("Config update error with code: \(error?.localizedDescription ?? "Unknown error")")
                return
            }
            print("Updated keys: \(configUpdate.updatedKeys)")
        }
    }

    func fetchAndActivateStringValue(configKey: String, completionHandler: @escaping (String?, Error?) -> Void) {
        Task {
            do {
                try await remoteConfig.fetchAndActivate()
                let result = remoteConfig[configKey].stringValue ?? ""
                completionHandler(result, nil)
            } catch {
                completionHandler(nil, error)
            }
        }
    }

    func fetchAndActivateValue(configKey: String, completionHandler: @escaping (KotlinFloat?, Error?) -> Void) {
        Task {
            do {
                try await remoteConfig.fetchAndActivate()
                let result = remoteConfig[configKey].numberValue.floatValue
                completionHandler(KotlinFloat(value: result), nil) // Wrap Float in KotlinFloat
            } catch {
                completionHandler(nil, error)
            }
        }
    }

    func fetchAndActivateBooleanValue(configKey: String, completionHandler: @escaping (KotlinBoolean?, Error?) -> Void) {
        Task {
            do {
                try await remoteConfig.fetchAndActivate()
                let result = remoteConfig[configKey].boolValue
                completionHandler(KotlinBoolean(value: result), nil) // Wrap Bool in KotlinBoolean
            } catch {
                completionHandler(nil, error)
            }
        }
    }
}
